scripts/gen_syscalls.py

```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2017 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+import re
+import argparse
+import os
+
+
+api_regex = re.compile(r'''
+__(syscall|syscall_inline)\s+   # __syscall or __syscall_inline
+static\s+inline\s+              # All prototypes are static inline functions
+([^(]+)                         # type and name of system call (split later)
+[(]                             # Function opening parenthesis
+([^)]*)                         # Arg list (split later)
```

Comment:
Note that this will fail if anything is parenthesized inside the function declaration, so no function pointers will work.  I don't know that that's a huge problem specifically, but long term it would be good to do proper parsing here.  The function type signatures are in the DWARF output, right?  Maybe you could try the same trickery here you did with the kobj stuff?  Or if not you could experiment with C++ name mangling (which I think can be turned on per-function using a GCC attribute, even in C code) to extract the same info.

Reply:
at the moment this doesn't cause issues as APIs which register callbacks are excluded from being added as system calls for security reasons, and APIs like k_thread_create() use a typedef for the entry point.

still, this is fragile. i'm spending some time today looking to see how far I can get with DWARF parsing.

Reply:
DWARF parsing is possible, but to get it completely right requires a hefty amount of code.
I propose we see how far we get in its current incarnation.
I'll work up a patch that will at least fail gracefully if a non-typedef'd function pointer is used.